Allow using different preference defaults depending on a Solidus version

waiting-for-dev opened the pull request:
This feature is similar to [Rail's
`load_defaults`](https://guides.rubyonrails.org/configuring.html#results-of-config-load-defaults).
However, instead of loading imperatively, the defaults for a given
version keep the history of the value in the preference declaration.
Both systems are upgrade-friendly, as users need to adjust the new
version defaults manually. The main advantage of this commit
implementation is communication (as the code tells about its current
state and history). However, it's architecturally more complex.

`Spree::Core::VersionedValue` implements the core behavior. This class
accepts a specification of how a value has changed in time and returns
the result for a given solidus version. The specification consists of an
initial value and a series of boundaries when it changed:

```ruby
value = Spree::Core::VersionedValue.new(false, "3.0" => true)
value.call("2.0") # => false
value.call("3.0") # => true
```

`Spree::AppConfiguration` bundles the behavior into the preferences
system. Its `.by_version` method builds a `Proc` that accepts a solidus
version and returns the corresponding value. It's meant to be used in
the `default:` keyword argument:

```ruby
preference :foo, :boolean, default: by_version(false, "3.0" => false)
```

Accordingly, `Spree::Preferences::Preferable` has been modified to
provide arguments to the `default:` `Proc`. As this module is not only
included in `Spree::AppConfiguration`, it defaults to supply no
arguments to it, but `Spree::AppConfiguration` overrides it to add a
`#loaded_defaults` attribute.

Users can specify the version defaults they want to use through the
`#load_defaults` method in `Spree::AppConfiguration` (which gets bound
in the `Spree.config` method used in the `spree` initializer):

```ruby
Spree.config do |config|
  config.load_defaults '3.1'
end
```

We have modified the `spree` generator template to add that line of code
defaulting to the current Solidus version. As the sibling step, we
should create an update task creating another initializer à la
`new_framework_defaults` in Rails.

The tricky part in the rollout of this new feature could be the first
upgrade after it, as current Solidus installations don't know about
`load_defaults`. Therefore, an upgrade will leave them with the new
defaults unless they manually add the `load_defaults` line.

Related to discussion: https://github.com/solidusio/solidus/discussions/4045

**Checklist:**
- [x] I have followed [Pull Request guidelines](https://github.com/solidusio/solidus/blob/master/CONTRIBUTING.md#pull-request-guidelines)
- [x] I have added a detailed description into each commit message
- [ ] I have updated Guides and README accordingly to this change (if needed)
- [x] I have added tests to cover this change (if needed)
- [ ] I have attached screenshots to this PR for visual changes (if needed)

aldesantis reviewed (approved):
@waiting-for-dev this is really impressive, thanks! I agree that it's architecturally more complex, but the implementation is simple and elegant enough to make dealing with that complexity almost a no-brainer.

kennyadsl commented:
Before merging, I'd like to understand how we want to communicate this to users. Can a section similar to this [Rails one](https://guides.rubyonrails.org/upgrading_ruby_on_rails.html#configure-framework-defaults) be enough?

kennyadsl commented on core/spec/lib/spree/app_configuration_spec.rb:
```diff
@@ -36,6 +36,18 @@
     expect(prefs.pricing_options_class).to eq Spree::Variant::PriceSelector.pricing_options_class
   end
 
+  it "allows defining diferent defaults depending on the Solidus version" do
```
small typo `diferent` -> `different`

waiting-for-dev replied:
:mag: fixed, thanks!

kennyadsl reviewed (approved):
Left a non-blocking comment. Thank you Marc, this is awesome!

waiting-for-dev commented:
I added a new commit generalizing the possibility to use different defaults by version to any `Preferable` class, and not only to `Spree::AppConfiguration`. Please, @kennyadsl , @aldesantis , I'd appreciate if you can take a look at the second commit.

waiting-for-dev commented:
> I added a new commit generalizing the possibility to use different defaults by version to any `Preferable` class, and not only to `Spree::AppConfiguration`. Please, @kennyadsl , @aldesantis , I'd appreciate if you can take a look at the second commit.

I pushed a new commit where, instead of enabling the new feature for any `Preferable` class, it's only done for descendants of `Spree::Preferences::Configuration`. I.e., configuration classes for core, backend, frontend & engine. After talking with @kennyadsl , we decided to set other `Preferable` classes aside, as they are ActiveRecord models that have very different behavior and can be configured through the backend.

aldesantis commented:
@waiting-for-dev makes sense to me!

jarednorman reviewed (approved):
A lot of moving parts here, but I've gone over it a couple times now and I think it all makes sense.

kennyadsl commented:
I think we are ready here, can we just rebase to remove commit 2 and 3 (3 reverts 2 and we can probably have the history cleaner by doing so).

waiting-for-dev commented:
> I think we are ready here, can we just rebase to remove commit 2 and 3 (3 reverts 2 and we can probably have the history cleaner by doing so).

I squashed all the commits and updated the commit message. I also rebased from master.

I made an extra addition, adding the `load_defaults` call in the backend, frontend & API configuration blocks in the `spree.rb` initializer template.

jarednorman reviewed (approved):
The approach for handling changes is very clever.

kennyadsl commented:
Merging, but I think we need to call `config.load_defaults` in the dummy apps as well.

waiting-for-dev commented:
> Merging, but I think we need to call `config.load_defaults` in the dummy apps as well.

Thanks, @kennyadsl. It's not needed because of `load_defaults` defaulting to the current Solidus version, but I'm introducing it on #4087 to avoid the warning we introduce there.
